File: scripts/ec2_backup.py
# # Automated EC2 backup script, that makes an instance image with AS Lambda.
# # It targets all instances with a tag "Backup: True"

import collections
import pprint

import boto3
import datetime

ec = boto3.client('ec2')


# Get ec2 client
def lambda_handler(event, context):
	reservations = ec.describe_instances(Filters=[
		{
			'Name': 'tag:Name', 'Values': ['Poc']
		},
	]).get('Reservations', [])
	instances = sum([[i for i in r['Instances']] for r in reservations], [])
	print("Found %d instances that need backing up" % len(instances))
	to_tag = collections.defaultdict(list)
	for instance in instances:
		try:
			retention_days = [
				int(t.get('Value')) for t in instance['Tags']
				if t['Key'] == 'Retention'
			][0]
			pprint.pprint(retention_days)
		except IndexError:
			retention_days = 7
			pprint.pprint(retention_days)
			# Create_image(instance_id, name, description=None, no_reboot=False, block_device_mapping=None, dry_run=False)
			# DryRun, InstanceId, Name, Description, NoReboot, BlockDeviceMappings
			for tag in instance['Tags']:
				if tag['Key'] == 'Name':
					instance_name = tag['Value']
			pprint.pprint(instance_name)

			create_time = datetime.datetime.now()
			create_fmt = create_time.strftime('%Y-%m-%d-%H%M')

			AMIid = ec.create_image(
				InstanceId=instance['InstanceId'],
				Name=instance_name + " - " + instance['InstanceId'] + " from " + create_fmt,
				Description="Lambda created AMI of instance " + instance['InstanceId'] + " from " + create_fmt,
				NoReboot=True,
				DryRun=False)

			pprint.pprint(instance)
			# # Can be used if NoReboot is set to False, and number of instances is small
			# # so that waiting will not go over Lambda timeout of 15 min
			# Wait for image to be available
			# waiter = ec.get_waiter('image_available')
			# waiter.wait(Filters=[{
			# 	'Name': 'image-id',
			# 	'Values': [AMIid['ImageId']]
			# }])
			to_tag[retention_days].append(AMIid['ImageId'])
			print("Retaining AMI %s of instance %s for %d days" % (
				AMIid['ImageId'],
				instance['InstanceId'],
				retention_days,
			))
			pprint.pprint(to_tag)

			pprint.pprint(to_tag.keys())
		for retention_days in to_tag.keys():
			delete_date = datetime.date.today() + datetime.timedelta(
				days=retention_days)
		delete_fmt = delete_date.strftime('%Y-%m-%d')
		print("Will delete %d AMIs on %s" % (len(to_tag[retention_days]), delete_fmt))

		# break

		ec.create_tags(Resources=to_tag[retention_days], Tags=[{'Key': 'DeleteOn', 'Value': delete_fmt}])

		pprint.pprint(to_tag[retention_days])

File: scripts/ec2_backup.go
// Added comment to test action
package main

import (
	"context"
	"fmt"
	"github.com/aws/aws-lambda-go/lambda"
	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/service/ec2"
	"github.com/aws/aws-sdk-go-v2/service/ec2/types"
	"github.com/aws/aws-sdk-go-v2/service/sns"
	"log"
	"os"
	"strconv"
	"time"
)

func main() {
	lambda.Start(HandleRequest)
}
func HandleRequest() {
	// Using the SDK's default configuration, loading additional config
	// and credentials values from the environment variables, shared
	// credentials, and shared configuration files
	cfg, err := config.LoadDefaultConfig(context.TODO())
	if err != nil {
		log.Fatalf("unable to load SDK config, %v", err)
	}
	ec2client := ec2.NewFromConfig(cfg)
	// define sns topic client
	snsClient := sns.NewFromConfig(cfg)
	// get sns topic arn from environment variable
	topicArn := os.Getenv("SNS_TOPIC_ARN")
	if topicArn == "" {
		log.Fatal("SNS_TOPIC_ARN environment variable must be set")
	}

	params := &ec2.DescribeInstancesInput{
		Filters: []types.Filter{
			{
				Name:   aws.String("tag:Backup"),
				Values: []string{"true"},
			},
		},
	}

	instances, err := ec2client.DescribeInstances(context.TODO(), params)
	if err != nil {
		log.Fatalf("unable to get instances, %v", err)
	}
	// iterate over reservations and print instance IDs and make instance snapshots
	for _, reservation := range instances.Reservations {
		for _, instance := range reservation.Instances {
			fmt.Println("Instance ID: ", *instance.InstanceId)
			// get backup retention days from tag values
			retentionDays := getBackupRetentionDays(instance)
			deleteDate := time.Now().AddDate(0, 0, retentionDays).Format("2006-01-02")
			err = makeInstanceSnapshot(context.TODO(), cfg, *instance.InstanceId, deleteDate)
			if err != nil {
				log.Fatalf("unable to make instance snapshot, %v", err)
			}
			// publish message to sns topic
			_, err := snsClient.Publish(context.TODO(), &sns.PublishInput{
				Subject:  aws.String("EC2 Snapshot notification"),
				Message:  aws.String("Instance ID: " + *instance.InstanceId + " Snapshot ID: " + *instance.InstanceId + "-" + time.Now().Format("2006-01-02")),
				TopicArn: aws.String(topicArn),
			})
			if err != nil {
				log.Fatalf("unable to publish message to sns topic, %v", err)
			}
		}
	}
}

// makeInstanceSnapshot creates an image of an instance and add tags of a date
func makeInstanceSnapshot(ctx context.Context, cfg aws.Config, instanceID string, delete string) error {
	svc := ec2.NewFromConfig(cfg)
	input := &ec2.CreateImageInput{
		Name:        aws.String(instanceID + "-" + time.Now().Format("2006-01-02")),
		Description: aws.String("Lambda Snapshot of " + instanceID),
		InstanceId:  aws.String(instanceID),
		TagSpecifications: []types.TagSpecification{
			{
				ResourceType: types.ResourceTypeImage,
				Tags: []types.Tag{
					{
						Key:   aws.String("DeleteOn"),
						Value: aws.String(delete),
					},
				},
			},
		},
	}
	result, err := svc.CreateImage(ctx, input)
	if err != nil {
		return err
	}
	fmt.Println("Snapshot ID: ", *result.ImageId)
	return nil
}

// getBackupRetentionDays will get the backup retention days from the tag value
// if the tag is not set, it will return 7 days
func getBackupRetentionDays(instance types.Instance) int {
	for _, tag := range instance.Tags {
		if *tag.Key == "BackupRetentionDays" {
			intValue, err := strconv.Atoi(*tag.Value)
			if err != nil {
				log.Fatalf("unable to convert tag value to int, %v", err)
				return 7
			}
			return intValue
		}
	}
	return 0
}
